src/XR/features/WebXRWalkingLocomotion.ts

```diff
@@ -0,0 +1,415 @@
+import { Vector2, Vector3 } from "../../Maths/math.vector";
+import { Observable } from "../../Misc/observable";
+import { Nullable } from "../../types";
+import { WebXRFeatureName, WebXRFeaturesManager } from "../webXRFeaturesManager";
+import { WebXRSessionManager } from "../webXRSessionManager";
+import { WebXRAbstractFeature } from "./WebXRAbstractFeature";
+
+class CircleBuffer {

```

Comment:
wondering - would it make sense to put all of those outside of the XR scope to allow reusing it in different scenarios?

Reply:
_Hmm_, not sure. Unless you feel strongly otherwise, I'd rather wait until we have occasion to use them somewhere else, at which point we can export them then. My reasons for this are twofold: (1) these were written as helpers for this particular scenario and are overwhelmingly specific to it, so at least in their current forms I don't actually anticipate them being well-suited to another task (with the possible exception of the `CircleBuffer`); and (2) given how experimental this is, I'm a bit leery of making it easy for others to take a dependency on the implementation details of this because I may want to completely change those implementation details if I think of a way to do this better later. For example, cross-correlation might be another way to tackle this same recognition problem; I didn't like it when I tried it before, but if I try it again and find a way to make it better, I'd like to be able to eliminate the current implementation details entirely in order to replace them with a superior alternative that operates completely differently.

Reply:
This is totally your decision. I just asked :-)